### PR TITLE
Optional class properties

### PR DESCRIPTION
**Pull request by ahejlsberg**
This PR makes it possible to declare optional properties and methods in classes, similar to what is already permitted in interfaces. For example:

```typescript
class Bar {
    a: number;
    b?: number;
    f() {
        return 1;
    }
    g?(): number;  // Body of optional method can be omitted
    h?() {
        return 2;
    }
}
```

When compiled in `--strictNullChecks` mode, optional properties and methods automatically have `undefined` included in their type. Thus, the `b` property above is of type `number | undefined` and the `g` method above is of type `(() => number) | undefined`. Type guards can be used to strip away the `undefined` part of the type:

```typescript
function test(x: Bar) {
    x.a;  // number
    x.b;  // number | undefined
    x.f;  // () => number
    x.g;  // (() => number) | undefined
    let f1 = x.f();            // number
    let g1 = x.g && x.g();     // number | undefined
    let g2 = x.g ? x.g() : 0;  // number
}
```

### CODE REVIEW

File: tests/baselines/reference/objectTypesWithOptionalProperties.errors.txt [hunk 1]
```diff
@@ -16,8 +14,6 @@ tests/cases/conformance/types/objectTypeLiteral/methodSignatures/objectTypesWith
     
     class C {
         x?: number; // error
```
Comment (mhegazy) [inline tests/baselines/reference/objectTypesWithOptionalProperties.errors.txt]: comments are out of date now.

**Comment by mhegazy** [comment]
can you add a declaration emit test.

**Comment by mhegazy** [comment]
can you also add a test with `--strictNullchecks` and an optional parameter property declaration, e.g. `constructor(private a?: number) { }`

**Comment by mhegazy** [comment]
also another one for extending classes, making optionals non-optional, and making non-optionals optional.

**Comment by CreepGin** [comment]
Just curious, can decorators fetch information that can determine whether the target is declared optional?

**Comment by mhegazy** [comment]
> can decorators fetch information that can determine whether the target is declared optional?

no, see https://github.com/Microsoft/TypeScript/issues/8126

**Comment by Roam-Cooper** [comment]
Optional class properties don't seem to work for abstract properties in typescript 2.0.3, subclasses of the abstract class are forced to implement the property.

**Comment by aluanhaddad** [comment]
@Roam-Cooper that's the point of the `abstract` modifier, to require that an implementation is provided by a subclass. If it's optional you don't have to override it meaning it should not be abstract.

**Comment by Roam-Cooper** [comment]
Ah, yes, wasn't thinking straight! My bad. 😃

**Comment by bradenhs** [comment]
Does this PR enable making class getters/setters optional? I can't figure out how to mark getters/setters as optional. I have no idea where the question mark would go.

**Comment by Venryx** [comment]
@bradenhs I have the same question. Would be nice for my situation, where I want to make a custom derived class (which has getters) compatible with the base class definition/type-shape.

**Comment by bradenhs** [comment]
@Venryx I made an issue for this (https://github.com/Microsoft/TypeScript/issues/14417) but it was shot down. I don't think the guy reviewing the issue fully understood what I was getting at. You could create another issue and give a better explanation of the problem though. If you do let me know!

**Pull request merged**